### PR TITLE
Fetch AWS region from EC2 instance metadata

### PR DESCRIPTION
**Pull request by stanislavb**
## Summary

sig4() and query() upon getting a location set to None will attempt to fetch it from EC2 instance metadata before falling back to us-east-1.
## Why not use salt.utils.iam?

salt.utils.s3 is the only file in salt repository that imports salt.utils.iam. With #26378 salt.utils.iam.get_iam_metadata is deprecated and this change opens up for deprecating salt.utils.iam.get_iam_region, solving it in a neater way in the more feature complete and more often used library.
## Change details in salt.utils.aws

get_region_from_metadata() uses caching just like creds() does. If it can't connect to the metadata service, it does not try again. If a value is successfully retrieved, it is cached so that future calls are faster.

get_location() will call get_region_from_metadata() if location is not explicitly set in query function call or salt config.

If sig4() receives a None location value, it tries to fetch it using get_region_from_metadata()

### CODE REVIEW

**Comment by jfindlay** [comment]
Awesome, thanks @stanislavb.

**Comment by techhat** [comment]
2015.5? Why go back so far?

**Comment by stanislavb** [comment]
@techhat The reasoning for developing 2015.5 is that 2015.5.3 is the currently available salt version in EPEL repositories for RedHat/CentOS 6 & 7 and also Amazon Linux. It is widely deployed at my workplace and probably many others.

The reason for developing salt.utils.aws starting with 2015.5 is that this lib is exactly the same in 2015.8 and therefore it is simple to improve both versions.

It might seem weird to introduce features in an old release, but the whole story is that I previously wrote salt.utils.iam.get_iam_region() before discovering the way more used salt.utils.aws and the method it uses to cache things to avoid unnecessary network calls. If this change makes sense to merge, the plan is to switch salt.utils.s3 to use this bit of code, in a separate pull request. :)

**Comment by cachedout** [comment]
Ping @techhat

**Comment by techhat** [comment]
@cachedout, I think it's a perfectly reasonable assessment.

File: salt/utils/aws.py [hunk 1]
```diff
@@ -442,15 +448,54 @@ def query(params=None, setname=None, requesturl=None, location=None,
     return ret
 
 
+def get_region_from_metadata():
+    '''
+    Try to get region from instance identity document and cache it
+    '''
```
Comment (rallytime) [inline salt/utils/aws.py]: Can you add a `.. versionadded:: 2015.5.5` tag to the docstring here?

**Comment by rallytime** [comment]
@stanislavb Thanks for this fix. This looks good and we'll be happy get this merged if you could add the versionadded tag that I mentioned above. Thanks!

**Comment by stanislavb** [comment]
@rallytime added versionadded tag and rebased on latest 2015.5

**Comment by rallytime** [comment]
Awesome! Thanks! I'll let these tests run and then we'll get this merged.

**Pull request merged**